File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/core/JavaScriptClassBuilder.kt
package ca.logaritm.dezel.core

import android.util.Log
import java.lang.reflect.Method

/**
 * @class JavaScriptClassBuilder
 * @since 0.1.0
 * @hidden
 */
internal class JavaScriptClassBuilder : JavaScriptBuilder() {

	companion object {

		/**
		 * @method build
		 * @since 0.1.0
		 * @hidden
		 */
		internal fun build(context: JavaScriptContext, template: Class<*>): JavaScriptValue {

			val statics = context.createEmptyObject()
			val prototype = context.createEmptyObject()

			forEach(template, fun(name: String, type: Type, method: Method) {

				if (type == Type.FUNCTION && name == "constructor") {
					this.createConstructor(context, prototype.handle, name, method.name, template)
					return
				}

				if (type == Type.FUNCTION) {
					this.createFunction(context, prototype.handle, name, method.name, template)
					return
				}

				if (type == Type.GETTER) {
					this.createGetter(context, prototype.handle, name, method.name, template)
					return
				}

				if (type == Type.SETTER) {
					this.createSetter(context, prototype.handle, name, method.name, template)
					return
				}

				if (type == Type.STATIC_FUNCTION) {
					this.createStaticFunction(context, prototype.handle, name, method.name, template)
					return
				}

				if (type == Type.STATIC_GETTER) {
					this.createStaticGetter(context, prototype.handle, name, method.name, template)
					return
				}

				if (type == Type.STATIC_SETTER) {
					this.createStaticSetter(context, prototype.handle, name, method.name, template)
					return
				}

			})

			val constructor = prototype.property("constructor")
			constructor.property("prototype", prototype)
			constructor.property("statics", statics)
			return constructor
		}

		/**
		 * @method createConstructor
		 * @since 0.1.0:
		 * @hidden
		 */
		private fun createConstructor(context: JavaScriptContext, target: Long, name: String, method: String, cls: Class<*>) {
			JavaScriptClassBuilderExternal.createConstructor(context.handle, target, name, method, cls, context)
		}

		/**
		 * @method createFunction
		 * @since 0.1.0:
		 * @hidden
		 */
		private fun createFunction(context: JavaScriptContext, target: Long, name: String, method: String, cls: Class<*>) {
			JavaScriptClassBuilderExternal.createFunction(context.handle, target, name, method, cls, context)
		}

		/**
		 * @method createGetter
		 * @since 0.1.0:
		 * @hidden
		 */
		private fun createGetter(context: JavaScriptContext, target: Long, name: String, method: String, cls: Class<*>) {
			JavaScriptClassBuilderExternal.createGetter(context.handle, target, name, method, cls, context)
		}

		/**
		 * @method createSetter
		 * @since 0.1.0:
		 * @hidden
		 */
		private fun createSetter(context: JavaScriptContext, target: Long, name: String, method: String, cls: Class<*>) {
			JavaScriptClassBuilderExternal.createSetter(context.handle, target, name, method, cls, context)
		}

		/**
		 * @method createStaticFunction
		 * @since 0.7.0:
		 * @hidden
		 */
		private fun createStaticFunction(context: JavaScriptContext, target: Long, name: String, method: String, cls: Class<*>) {
			JavaScriptClassBuilderExternal.createStaticFunction(context.handle, target, name, method, cls, context)
		}

		/**
		 * @method createStaticGetter
		 * @since 0.7.0:
		 * @hidden
		 */
		private fun createStaticGetter(context: JavaScriptContext, target: Long, name: String, method: String, cls: Class<*>) {
			JavaScriptClassBuilderExternal.createStaticGetter(context.handle, target, name, method, cls, context)
		}

		/**
		 * @method createStaticSetter
		 * @since 0.7.0:
		 * @hidden
		 */
		private fun createStaticSetter(context: JavaScriptContext, target: Long, name: String, method: String, cls: Class<*>) {
			JavaScriptClassBuilderExternal.createStaticSetter(context.handle, target, name, method, cls, context)
		}
	}
}

File: platforms/ios/Dezel/Dezel/Core/JavaScriptClassBuilder.swift
/**
 * @class JavaScriptClassBuilder
 * @super JavaScriptBuilder
 * @since 0.1.0
 * @hidden
 */
internal final class JavaScriptClassBuilder: JavaScriptBuilder {

	//--------------------------------------------------------------------------
	// MARK: Methods
	//--------------------------------------------------------------------------

	/**
     * @method build
	 * @since 0.1.0
	 * @hidden
     */
	internal class func build(_ context: JavaScriptContext, template: AnyClass) -> JavaScriptValue {

		let prototype = context.createEmptyObject()
		let statics = context.createEmptyObject()

		JavaScriptBuilder.forEach(template, callback: { (name, type, sel, imp) -> Void in

			if (type == .function && name == "constructor") {
				JavaScriptValueDefineConstructor(context.handle, prototype.handle, JavaScriptClassConstructorWrapper(context: context, cls: template, sel: sel, imp: imp, name: String(describing: template)).function)
				return
			}

			if (type == .function) {
				JavaScriptValueDefineFunction(context.handle, prototype.handle, name, JavaScriptClassFunctionWrapper(context: context, cls: template, sel: sel, imp: imp, name: name).function)
				return
			}

			if (type == .getter) {
				JavaScriptValueDefinePropertyGetter(context.handle, prototype.handle, name, JavaScriptClassGetterWrapper(context: context, cls: template, sel: sel, imp: imp, name: name).function)
				return
			}

			if (type == .setter) {
				JavaScriptValueDefinePropertySetter(context.handle, prototype.handle, name, JavaScriptClassSetterWrapper(context: context, cls: template, sel: sel, imp: imp, name: name).function)
				return
			}

			if (type == .staticFunction) {
				JavaScriptValueDefineFunction(context.handle, statics.handle, name, JavaScriptClassStaticFunctionWrapper(context: context, cls: template, sel: sel, imp: imp, name: name).function)
				return
			}

			if (type == .staticGetter) {
				JavaScriptValueDefinePropertyGetter(context.handle, statics.handle, name, JavaScriptClassStaticGetterWrapper(context: context, cls: template, sel: sel, imp: imp, name: name).function)
				return
			}

			if (type == .staticSetter) {
				JavaScriptValueDefinePropertySetter(context.handle, statics.handle, name, JavaScriptClassStaticSetterWrapper(context: context, cls: template, sel: sel, imp: imp, name: name).function)
				return
			}
		})

		let constructor = prototype.property("constructor")
		constructor.property("prototype", value: prototype)
		constructor.property("statics", value: statics)

		return constructor
	}
}
